pkg/parse/run.go

```diff
@@ -201,91 +128,179 @@ func DefaultRunFunc(ctx context.Context, r Reconciler, trigger string) RunResult
+	// Skip parse-apply-watch if the trigger is `triggerSync` (aka "reimport")
+	// and there are no new source changes. The reasons are:
+	//   * If a former parse-apply-watch sequence for syncPath succeeded, there is no need to run the sequence again;
 	//   * If all the former parse-apply-watch sequences for syncDir failed, the next retry will call the sequence.
-	if trigger == triggerSync && oldSyncDir == newSyncDir {
+	if trigger == triggerSync && oldSyncPath == newSyncPath {
 		return result
 	}
 
-	errs := r.ParseAndUpdate(ctx, trigger)
-	if errs != nil {
-		state.invalidate(errs)
+	if errs := r.ParseAndUpdate(ctx, trigger); errs != nil {
+		state.RecordFailure(opts.Clock, errs)
 		return result
 	}
 
 	// Only checkpoint the state after *everything* succeeded, including status update.
-	state.checkpoint()
+	state.RecordSyncSuccess(opts.Clock)
 	result.Success = true
 	return result
 }
 
+// Fetch waits for the *-sync sidecars to fetch the source manifests to the
+// shared source volume.
+// Updates the RSync status (source status and syncing condition).
+func (r *reconciler) Fetch(ctx context.Context) (*SourceStatus, cmpath.Absolute, status.MultiError) {
+	opts := r.Options()
+	state := r.ReconcilerState()
+	var syncPath cmpath.Absolute
+	newSourceStatus := &SourceStatus{}
+
+	// pull the source commit and directory with retries within 5 minutes.
+	newSourceStatus.Commit, syncPath, newSourceStatus.Errs = hydrate.SourceCommitAndDirWithRetry(util.SourceRetryBackoff, opts.SourceType, opts.SourceDir, opts.SyncDir, opts.ReconcilerName)
+
+	// Add pre-sync annotations to the object.
+	// If updating the object fails, it's likely due to a signature verification error
+	// from the webhook. In this case, add the error as a source error.
+	if newSourceStatus.Errs == nil {
+		if err := r.SyncStatusClient().SetImageToSyncAnnotation(ctx, newSourceStatus.Commit); err != nil {
+			newSourceStatus.Errs = status.Append(newSourceStatus.Errs, err)
+		}
+	}
+
+	// Generate source spec from Reconciler config
+	newSourceStatus.Spec = SourceSpecFromFileSource(opts.FileSource, opts.SourceType, newSourceStatus.Commit)
+
+	// Only update the source status if there are errors or the commit changed.
+	// Otherwise, parsing errors may be overwritten.
+	// TODO: Decouple fetch & parse stages to use different status fields
+	if newSourceStatus.Errs != nil || state.status.SourceStatus == nil || newSourceStatus.Commit != state.status.SourceStatus.Commit {
+		newSourceStatus.LastUpdate = nowMeta(opts.Clock)
+		if state.status.needToSetSourceStatus(newSourceStatus) {
+			klog.V(3).Info("Updating source status (after fetch)")
+			if statusErr := r.SyncStatusClient().SetSourceStatus(ctx, newSourceStatus); statusErr != nil {
+				return newSourceStatus, syncPath, status.Append(newSourceStatus.Errs, statusErr)
+			}
+			state.status.RecordSourceStatusUpdate(newSourceStatus)
```

Comment:
I agree that this is suboptimal, but the problem isn't that lower level functions are managing the state, but actually that this layer of abstraction needs to track state at all.

If this were a standard controller, the state would be stored in the RSync object itself and the whole RSync object would be cached. So you would just ask the object cache what the latest RSync status is, rather than storing the state in memory in some similar but not quite identical struct. Then you would just pass the latest copy of the RSync down to the inner methods and they could check it directly instead of hitting the cache again.

But we got here with organic growth, rather than intentional design changes. So we can't fix them all at once.

This particular smell I find less pressing than reducing the event/timer loop down to a single event type. Once we get that squared away, it may be more obvious how to simplify the in-memory cache.